python/tests/bro.py

```diff
@@ -0,0 +1,151 @@
+#! /usr/bin/env python
+"""bro %s -- compression/decompression utility using the Brotli algorithm."""
+
+from __future__ import print_function
+import brotli
+import getopt
+import sys
+import os
+
+
+__usage__ = """\
+Usage: bro [--force] [--decompress] [--input filename] [--output filename]
+    [--mode 'text'|'font'] [--transform] [--bufsize]"""
+
+__version__ = '0.1'
+
+
+BROTLI_MODES = {
+    'text': brotli.MODE_TEXT,
+    'font': brotli.MODE_FONT
+}
+
+
+def get_binary_stdout():
+    if sys.version_info[0] < 3:
+        if sys.platform == 'win32':
+            # set stdout binary flag on python2.x (Windows)
+            import msvcrt
+            msvcrt.setmode(sys.stdout.fileno(), os.O_BINARY)
+        return sys.stdout
+    else:
+        # use 'buffer' attribute to write binary to stdout on python3.x
+        if hasattr(sys.stdout, 'buffer'):
+            return sys.stdout.buffer
+        else:
+            return sys.__stdout__.buffer
+
+
+def get_binary_stdin():
+    if sys.version_info[0] < 3:
+        if sys.platform == 'win32':
+            # set stdin binary flag on python2.x (Windows)
+            import msvcrt
+            msvcrt.setmode(sys.stdin.fileno(), os.O_BINARY)
+        return sys.stdin
+    else:
+        # use 'buffer' attribute to read binary from stdin on python3.x
+        if hasattr(sys.stdin, 'buffer'):
+            return sys.stdin.buffer
+        else:
+            return sys.__stdin__.buffer
+
+
+def main(args):
+
+    options = parse_options(args)
+
+    if options.infile:
+        if not os.path.isfile(options.infile):
+            print('file "%s" not found' % options.infile, file=sys.stderr)
+            sys.exit(1)
+        with open(options.infile, "rb") as infile:
+            data = infile.read()
+    else:
+        if sys.stdin.isatty():
+            # interactive console, just quit
+            usage()
+        infile = get_binary_stdin()
+        data = infile.read()
+
+    if options.outfile:
+        if os.path.isfile(options.outfile) and not options.force:
+            print('output file exists')
+            sys.exit(1)
+        outfile = open(options.outfile, "wb")
+    else:
+        outfile = get_binary_stdout()
+
+    try:
+        if options.decompress:
+            bufsize = options.bufsize or 10*len(data)
```

Comment:
The default here seems a bit fragile. The compression ratio could easily be more that 10x, the easiest example is a huge file with only zeros in it.

If you want to be more robust, the solutions I can think of, in order of increasing complexity:
  1) Do the decompression in a loop, with exponentially increasing buffer sizes, until a limit is reached. The limit can be a command-line argument with a reasonable default, e.g. 1GB
  2) Add a python binding for the BrotliDecompressedSize()  and try to use that. Drawback is that it does not work if there are many metablocks, so you would have to fall back to solution 1)
  3) Add a python binding for the decompression function that uses two callbacks, and then you can allocate more output if it is needed.